source/tests/system/nidaqmx_driver_api_tests.cpp

```diff
@@ -2065,6 +2159,34 @@ TEST_F(NiDAQmxDriverApiTests, LoadedVoltageTask_ReadAIData_ReturnsDataInExpected
   EXPECT_DATA_IN_RANGE(read_response.read_array(), AI_MIN, AI_MAX);
 }
 
+TEST_F(NiDAQmxDriverApiTests, PerformBridgeShuntCalEx_Succeeds)
```

Comment:
Test failures:

```
[ RUN      ] NiDAQmxDriverApiTests.PerformBridgeShuntCalEx_Succeeds
unknown file: Failure
C++ exception with description "Selected physical channel does not support the measurement type required by the virtual channel you are creating.
Create a channel of a measurement type that is supported by the physical channel, or select a physical channel that supports the measurement type.
Property: DAQmx_AI_MeasType
Requested Value: DAQmx_Val_Bridge
Possible Values: DAQmx_Val_Current, DAQmx_Val_Resistance, DAQmx_Val_Strain_Gage, DAQmx_Val_Temp_BuiltInSensor, DAQmx_Val_Temp_RTD, DAQmx_Val_Temp_Thrmstr, DAQmx_Val_Temp_TC, DAQmx_Val_Voltage, DAQmx_Val_Voltage_CustomWithExcitation

Task Name: _unnamedTask<3C>

Status Code: -200431" thrown in the test body.
[  FAILED  ] NiDAQmxDriverApiTests.PerformBridgeShuntCalEx_Succeeds (5 ms)
[ RUN      ] NiDAQmxDriverApiTests.PerformStrainShuntCalEx_Succeeds
unknown file: Failure
C++ exception with description "Device to which the sensor is attached does not have an available internal excitation source.
Select another device with an available internal excitation source or supply external excitation.
Channel Name: ai0
Device: gRPCSystemTestDAQ

Task Name: _unnamedTask<3D>

Status Code: -200161" thrown in the test body.
[  FAILED  ] NiDAQmxDriverApiTests.PerformStrainShuntCalEx_Succeeds (3 ms)
```

"grpcbot", the VM we use to run these system tests, has a PXIe-6341 installed. This device doesn't support the Bridge (V/V) AI measurement type. It does support strain gage channels, as long as you configure them for external excitation. I don't know if it can successfully run a shunt calibration without additional signal conditioning hardware (SCC or SCXI).

I recommend trying the LabVIEW strain gage shipping example with a simulated X Series to see if you are going to be successful testing the shunt calibration API with a PXIe-6341. I don't know if it's worthwhile to reconfigure grpcbot to support more simulated DAQmx devices.

Reply:
@bkeryan I've tried it on LabVIEW, seems like PXIe-6341 does support strain gage channel with external excitation but doesn't support shunt calibration procedures. I did tried to set shuntResistorSource to External as well, and it still couldn't work. 

Since the configured grpcbot simulated device (PXIe-6341) doesn't suport Bridge AI measurement type and also strain gage shunt calibration, how do we reconfigure grpcbot to support more DAQmx devices to run test for these APIs?

Reply:
To do that, you would have to:
- Document which simulated devices are needed to run tests.
- Update https://github.com/ni/grpc-device/blob/main/source/tests/assets/grpc-device-daq-tests.nce . Consider converting it to a `.ini` file so it's human readable.
- Update grpcbot to create the simulated devices:
  - You could add them to nisimdev_simulate.sls, but this duplicates data from grpc-device-daq-tests.nce
  - You could update grpcbot to install & run nidaqmxconfig, like we do for nidaqmx-python's "nidaqmxbot"
  - After updating the config, run the grpcbot build/deploy pipeline

Ultimately, I am not convinced this is worthwhile for the amount of code it would be testing. I think it would be acceptable to either test that strain returns the correct error for an X Series or that skipUnsupportedChannels can skip all of the channels.